File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第3章FBI树.cpp
//FBI 树  lanqiaoOJ 571 

#include <bits/stdc++.h>
using namespace std;
char s[1100],tree[4400];             //tree[]存满二叉树
int ls(int p){ return p<<1;  }       //定位左儿子：p*2
int rs(int p){ return p<<1|1;}       //定位右儿子：p*2 + 1
void build_FBI(int p,int left,int right){
	if(left==right){           //到达叶子结点
        if(s[right]=='1')  tree[p]='I';
        else               tree[p]='B';
        return;
	}
	int mid=(left+right)/2;         //分成两半
	build_FBI(ls(p),left,mid);      //递归左半
	build_FBI(rs(p),mid+1,right);   //递归右半

	if(tree[ls(p)]=='B' && tree[rs(p)]=='B')    //左右儿子都是B，自己也是B
        tree[p]='B';
	else if(tree[ls(p)]=='I'&&tree[rs(p)]=='I') //左右儿子都是I，自己也是I
        tree[p]='I';
	else tree[p]='F';
}
void postorder (int p){   //后序遍历
	if(tree[ls(p)])	 postorder (ls(p));
	if(tree[rs(p)])	 postorder (rs(p));
	printf("%c",tree[p]);
}
int main(){
	int n;	scanf("%d",&n);
	scanf("%s",s+1);
	build_FBI(1,1,strlen(s+1));
	postorder(1);
}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第3章FBI树.py
#FBI 树 lanqiaoOJ 571

def build_FBI(p,left,right):
  if left==right:
    if s[right]=='1': tree[p]='I'
    else:  tree[p]='B'
    return
  mid=(left+right)//2
  build_FBI(2*p,left,mid)
  build_FBI(2*p+1,mid+1,right)

  if tree[2*p]=='B' and tree[2*p+1]=='B':  tree[p]='B'
  elif tree[2*p] == 'I' and tree[2*p+1]=='I': tree[p]='I'
  else: tree[p]='F'

def postorder(p):
  if tree[2*p] != '':  postorder(2*p)
  if tree[2*p+1] != '':   postorder(2*p+1)
  print(tree[p],end='')

n = int(input())
s = input()
tree=['']*4400
build_FBI(1,0, len(s)-1)
postorder(1)
